File: DSA/Trees/bst.py
class Node:
    def __init__(self, value):
        self.left: Node | None = None
        self.right: Node | None = None
        self.value = value

    def __str__(self) -> str:
        return str(self.value)


class BST:
    def __init__(self):
        self.root: Node | None = None

    def empty(self) -> bool:
        return self.root is None

    def insert(self, node: Node) -> None:
        if self.empty():
            self.root = node
        else:
            current = self.root
            while True:
                if node.value < current.value:
                    if current.left is None:
                        current.left = node
                        break
                    current = current.left
                elif node.value > current.value:
                    if current.right is None:
                        current.right = node
                        break
                    current = current.right
                else:
                    print("> Value already exists! ")
                    break

    def insert_many(self, vals: list[int]) -> None:
        for i in vals:
            self.insert(Node(i))

    def delete(self, value):
        def delete_node(node, value):
            if node is None:
                return node

            if value < node.value:
                node.left = delete_node(node.left, value)
            elif value > node.value:
                node.right = delete_node(node.right, value)
            else:
                if node.left is None:
                    return node.right
                elif node.right is None:
                    return node.left

                node.value = self.get_min(node.right)
                node.right = delete_node(node.right, node.value)

            return node

        self.root = delete_node(self.root, value)

    def get_max(self, node: Node) -> int:
        if node is None:
            return self._get_max(self.root)
        return self._get_max(node)

    def _get_max(self, node: Node) -> int:
        if node is None:
            return None

        while node.right is not None:
            node = node.right

        return node.value

    def get_min(self, node: Node) -> int:
        if node is None:
            return self._get_min(self.root)
        return self._get_min(node)

    def _get_min(self, node: Node) -> int:
        if node is None:
            return None

        while node.left is not None:
            node = node.left

        return node.value

    def find(self, value) -> bool:
        root = self.root
        while root != None and root.value != value:
            if value < root.value:
                root = root.left
            else:
                root = root.right
        if root == None:
            return False
        else:
            return True

    def find_node(self, value) -> None:
        res = self.find(value)
        if not res:
            print(f"> Not Found {value}")
        else:
            print(f"> Found {value}")

    def display(self):
        def in_order_traversal(node):
            if node:
                in_order_traversal(node.left)
                print(node.value, end=" ")
                in_order_traversal(node.right)

        def pre_order_traversal(node):
            if node:
                print(node.value, end=" ")
                in_order_traversal(node.left)
                in_order_traversal(node.right)

        def post_order_traversal(node):
            if node:
                in_order_traversal(node.left)
                in_order_traversal(node.right)
                print(node.value, end=" ")

        in_order_traversal(self.root)
        print("--> Inorder Traversal")
        pre_order_traversal(self.root)
        print("--> PreOrder Traversal")
        post_order_traversal(self.root)
        print("--> PostOrder Traversal")


# Credits: https://stackoverflow.com/a/65865825
def print_tree(root, value="value", left="left", right="right"):
    print("\n> BINARY TREE")

    def display(root, value=value, left=left, right=right):
        """Returns list of strings, width, height, and horizontal coordinate of the root."""
        # No child.
        if getattr(root, right) is None and getattr(root, left) is None:
            line = "%s" % getattr(root, value)
            width = len(line)
            height = 1
            middle = width // 2
            return [line], width, height, middle

        # Only left child.
        if getattr(root, right) is None:
            lines, n, p, x = display(getattr(root, left))
            s = "%s" % getattr(root, value)
            u = len(s)
            first_line = (x + 1) * " " + (n - x - 1) * "_" + s
            second_line = x * " " + "/" + (n - x - 1 + u) * " "
            shifted_lines = [line + u * " " for line in lines]
            return [first_line, second_line] + shifted_lines, n + u, p + 2, n + u // 2

        # Only right child.
        if getattr(root, left) is None:
            lines, n, p, x = display(getattr(root, right))
            s = "%s" % getattr(root, value)
            u = len(s)
            first_line = s + x * "_" + (n - x) * " "
            second_line = (u + x) * " " + "\\" + (n - x - 1) * " "
            shifted_lines = [u * " " + line for line in lines]
            return [first_line, second_line] + shifted_lines, n + u, p + 2, u // 2

        # Two children.
        left, n, p, x = display(getattr(root, left))
        right, m, q, y = display(getattr(root, right))
        s = "%s" % getattr(root, value)
        u = len(s)
        first_line = (x + 1) * " " + (n - x - 1) * "_" + s + y * "_" + (m - y) * " "
        second_line = (
            x * " " + "/" + (n - x - 1 + u + y) * " " + "\\" + (m - y - 1) * " "
        )
        if p < q:
            left += [n * " "] * (q - p)
        elif q < p:
            right += [m * " "] * (p - q)
        zipped_lines = zip(left, right)
        lines = [first_line, second_line] + [a + u * " " + b for a, b in zipped_lines]
        return lines, n + m + u, max(p, q) + 2, n + u // 2

    lines, *_ = display(root, value, left, right)
    for line in lines:
        print(line)


bst = BST()
bst.insert_many([5, 3, 8, 2, 1, 4, 6, 7, 9, 10])
print_tree(bst.root)
bst.display()

bst.delete(5)
print_tree(bst.root)
bst.display()

File: DSA/Trees/bst.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct TreeNode {
    int data;
    struct TreeNode *left;
    struct TreeNode *right;
} TreeNode;

// Function for creating a new BST node
TreeNode *createNode(int data) {
    TreeNode *node = (TreeNode *)malloc(sizeof(TreeNode));
    node->data = data;
    node->left = NULL;
    node->right = NULL;
}

// Function to insert a new value into the BST
TreeNode *insert(TreeNode *root, int data) {
    if (root == NULL) {
        printf("\n> Inserted %d\n", data);
        return createNode(data);
    }

    if (data < root->data) {
        root->left = insert(root->left, data);
    } else if (data > root->data) {
        root->right = insert(root->right, data);
    } else {
        printf("\n> Data Already Exists!\n");
        return root;
    }
    return root;
}

// Function to display the BST in-order
void display_in(TreeNode *root) {
    if (root != NULL) {
        display_in(root->left);
        printf("%d ", root->data);
        display_in(root->right);
    }
}

// Function to display the BST pre-order
void display_pre(TreeNode *root) {
    if (root != NULL) {
        printf("%d ", root->data);
        display_pre(root->left);
        display_pre(root->right);
    }
}

// Function to display the BST post-order
void display_post(TreeNode *root) {
    if (root != NULL) {
        display_post(root->left);
        display_post(root->right);
        printf("%d ", root->data);
    }
}

// Function to display the BST's min value node
TreeNode *minValueNode(TreeNode *root) {
    if (root->left != NULL) {
        return minValueNode(root->left);
    } else {
        return root;
    }
}

// Function to display the BST's max value node
TreeNode *maxValueNode(TreeNode *root) {
    if (root->right != NULL) {
        return maxValueNode(root->right);
    } else {
        return root;
    }
}

// Function to find a node in the BST
TreeNode *findNode(struct TreeNode *root, int data) {
    TreeNode *temp = root;
    while (temp != NULL && temp->data != data) {
        if (data < temp->data) {
            temp = temp->left;
        } else {
            temp = temp->right;
        }
    }
    if (temp == NULL) {
        printf("\n> Node Not Found! Returning NULL\n");
        return NULL;
    } else {
        printf("\n> Node Found! Returning Node\n");
        return temp;
    }
}

// Function to delete a node with given value from the BST
TreeNode *delete(TreeNode *root, int data) {
    if (root == NULL)
        return NULL;
    if (root->data < data)
        root->right = delete (root->right, data);
    else if (root->data > data)
        root->left = delete (root->left, data);
    else {
        // Leaf Node
        if (root->left == NULL && root->right == NULL) {
            free(root);
            return NULL;
        } // One Child: Right
        else if (root->left == NULL) {
            TreeNode *right = root->right;
            free(root);
            return right;
        } // One Child: Left
        else if (root->right == NULL) {
            TreeNode *left = root->left;
            free(root);
            return left;
        } // Both Children Present
        else {
            TreeNode *right_min_node = minValueNode(root->right);
            int right_min = right_min_node->data;
            root->data = right_min;
            delete (root->right, right_min);
        }
    }
    return root;
}

// Function to display (L D R) of a Node
void nodeDetails(TreeNode *root) {
    if (root == NULL) {
        printf("\n> ---- Node ----\n");
        printf("> Left Data Right\n");
        printf("> NULL NULL  NULL\n");

    } else {
        printf("\n> ---- Node ----\n");
        printf("> Left Data Right\n");
        
        if (root->left == NULL) {
            printf("> NULL ");
        } else {
            printf("> %4d ", root->left->data);
        }

        printf("%4d", root->data);

        if (root->right == NULL) {
            printf("  NULL \n");
        } else {
            printf(" %5d \n", root->right->data);
        }
    }
}

void main() {
    TreeNode *root = createNode(5);
    root = insert(root, 2);
    root = insert(root, 3);
    root = insert(root, 4);
    root = insert(root, 7);
    root = insert(root, 1);
    root = insert(root, 6);

    printf("\nTraversals\n");
    display_in(root);
    printf(" <--- InOrder Traversal\n");
    display_pre(root);
    printf(" <--- PreOrder Traversal\n");
    display_post(root);
    printf(" <--- PostOrder Traversal\n");

    TreeNode *min = minValueNode(root);
    nodeDetails(min);
    TreeNode *max = maxValueNode(root);
    nodeDetails(max);

    TreeNode *find = findNode(root, 5);
    nodeDetails(find);

    find = findNode(root, 11);
    nodeDetails(find);

    root = delete (root, 5);
    printf("\nTraversals\n");
    display_in(root);
    printf(" <--- InOrder Traversal\n");
    display_pre(root);
    printf(" <--- PreOrder Traversal\n");
    display_post(root);
    printf(" <--- PostOrder Traversal\n");
}
